Persistent volume is mandatory

I'm trying to start a Redis service without any persistent volume (ideal for testing...) and I get the following error in the container.

The CR created is:


```
apiVersion: redis.redis.opstreelabs.in/v1beta1
kind: Redis
metadata:
  name: service-redis
spec:
  kubernetesConfig:
    image: quay.io/opstree/redis:v6.2
    imagePullPolicy: IfNotPresent
    redisSecret:
      key: password
      name: redis-secret
    resources:
      limits:
        cpu: 101m
        memory: 128Mi
      requests:
        cpu: 101m
        memory: 128Mi
    serviceType: LoadBalancer
  redisConfig: {}
  redisExporter:
    enabled: true
    image: quay.io/opstree/redis-exporter:1.0
    imagePullPolicy: Always
    resources:
      limits:
        cpu: 100m
        memory: 128Mi
      requests:
        cpu: 100m
        memory: 128Mi
```


```
E0707 14:45:05.674354       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 420 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1670520, 0x23de090)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:74 +0xa6
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:48 +0x89
panic(0x1670520, 0x23de090)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
redis-operator/k8sutils.generateRedisStandaloneParams(...)
	/workspace/k8sutils/redis-standalone.go:59
redis-operator/k8sutils.CreateStandAloneRedis(0xc000810000, 0xc000810000, 0x1a4fc20)
	/workspace/k8sutils/redis-standalone.go:40 +0x5a0
redis-operator/controllers.(*RedisReconciler).Reconcile(0xc00052bb60, 0x1a2b5a0, 0xc0005aa630, 0xc00093add0, 0x4, 0xc00093adc0, 0xd, 0xc0005aa630, 0x40a27f, 0xc000030000, ...)
	/workspace/controllers/redis_controller.go:58 +0x328
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001a6a00, 0x1a2b4e0, 0xc00054c000, 0x16c7a60, 0xc000a4c9a0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:263 +0x317
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001a6a00, 0x1a2b4e0, 0xc00054c000, 0x203000)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:235 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1(0x1a2b4e0, 0xc00054c000)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:198 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000505f50)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc001167f50, 0x19f4d40, 0xc000a42b10, 0xc00054c001, 0xc000556000)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000505f50, 0x3b9aca00, 0x0, 0x1, 0xc000556000)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x1a2b4e0, 0xc00054c000, 0xc0007e0070, 0x3b9aca00, 0x0, 0x1)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x1a2b4e0, 0xc00054c000, 0xc0007e0070, 0x3b9aca00)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:195 +0x4e7
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x152ba00]

goroutine 420 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:55 +0x10c
panic(0x1670520, 0x23de090)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
redis-operator/k8sutils.generateRedisStandaloneParams(...)
	/workspace/k8sutils/redis-standalone.go:59
redis-operator/k8sutils.CreateStandAloneRedis(0xc000810000, 0xc000810000, 0x1a4fc20)
	/workspace/k8sutils/redis-standalone.go:40 +0x5a0
redis-operator/controllers.(*RedisReconciler).Reconcile(0xc00052bb60, 0x1a2b5a0, 0xc0005aa630, 0xc00093add0, 0x4, 0xc00093adc0, 0xd, 0xc0005aa630, 0x40a27f, 0xc000030000, ...)
	/workspace/controllers/redis_controller.go:58 +0x328
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001a6a00, 0x1a2b4e0, 0xc00054c000, 0x16c7a60, 0xc000a4c9a0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:263 +0x317
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001a6a00, 0x1a2b4e0, 0xc00054c000, 0x203000)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:235 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1(0x1a2b4e0, 0xc00054c000)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:198 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000505f50)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc001167f50, 0x19f4d40, 0xc000a42b10, 0xc00054c001, 0xc000556000)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000505f50, 0x3b9aca00, 0x0, 0x1, 0xc000556000)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x1a2b4e0, 0xc00054c000, 0xc0007e0070, 0x3b9aca00, 0x0, 0x1)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x1a2b4e0, 0xc00054c000, 0xc0007e0070, 0x3b9aca00)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:195 +0x4e7
```

--- a/k8sutils/redis-standalone.go
+++ b/k8sutils/redis-standalone.go
@@ -48,16 +48,19 @@ func CreateStandAloneRedis(cr *redisv1beta1.Redis) error {
 // generateRedisStandalone generates Redis standalone information
 func generateRedisStandaloneParams(cr *redisv1beta1.Redis) statefulSetParameters {
 	replicas := int32(1)
-	return statefulSetParameters{
-		Replicas:              &replicas,
-		NodeSelector:          cr.Spec.NodeSelector,
-		SecurityContext:       cr.Spec.SecurityContext,
-		PriorityClassName:     cr.Spec.PriorityClassName,
-		Affinity:              cr.Spec.Affinity,
-		Tolerations:           cr.Spec.Tolerations,
-		EnableMetrics:         cr.Spec.RedisExporter.Enabled,
-		PersistentVolumeClaim: cr.Spec.Storage.VolumeClaimTemplate,
+	res := statefulSetParameters{
+		Replicas:          &replicas,
+		NodeSelector:      cr.Spec.NodeSelector,
+		SecurityContext:   cr.Spec.SecurityContext,
+		PriorityClassName: cr.Spec.PriorityClassName,
+		Affinity:          cr.Spec.Affinity,
+		Tolerations:       cr.Spec.Tolerations,
+		EnableMetrics:     cr.Spec.RedisExporter.Enabled,
 	}
+	if cr.Spec.Storage != nil {
+		res.PersistentVolumeClaim = cr.Spec.Storage.VolumeClaimTemplate
+	}
+	return res
 }
 
 // generateRedisStandaloneContainerParams generates Redis container information

--- a/k8sutils/statefulset.go
+++ b/k8sutils/statefulset.go
@@ -2,6 +2,7 @@ package k8sutils
 
 import (
 	"context"
+
 	"github.com/banzaicloud/k8s-objectmatcher/patch"
 	"github.com/go-logr/logr"
 	appsv1 "k8s.io/api/apps/v1"
@@ -112,7 +113,7 @@ func generateStateFulSetsDef(stsMeta metav1.ObjectMeta, labels map[string]string
 	if params.Tolerations != nil {
 		statefulset.Spec.Template.Spec.Tolerations = *params.Tolerations
 	}
-	if *containerParams.PersistenceEnabled {
+	if containerParams.PersistenceEnabled != nil && *containerParams.PersistenceEnabled {
 		statefulset.Spec.VolumeClaimTemplates = append(statefulset.Spec.VolumeClaimTemplates, createPVCTemplate(stsMeta.Name, params.PersistentVolumeClaim))
 	}
 	AddOwnerRefToObject(statefulset, ownerDef)
@@ -168,7 +169,7 @@ func enableRedisMonitoring(params containerParameters) corev1.Container {
 // getVolumeMount gives information about persistence mount
 func getVolumeMount(name string, persistenceEnabled *bool) []corev1.VolumeMount {
 	var VolumeMounts []corev1.VolumeMount
-	if *persistenceEnabled && persistenceEnabled != nil {
+	if persistenceEnabled != nil && *persistenceEnabled {
 		VolumeMounts = []corev1.VolumeMount{
 			{
 				Name:      name,
@@ -218,7 +219,7 @@ func getEnvironmentVariables(role string, enabledPassword *bool, secretName *str
 			},
 		})
 	}
-	if *persistenceEnabled && persistenceEnabled != nil {
+	if persistenceEnabled != nil && *persistenceEnabled {
 		envVars = append(envVars, corev1.EnvVar{Name: "PERSISTENCE_ENABLED", Value: "true"})
 	}
 	return envVars

